File: apps/api/app/schemas/inspection.py
"""Pydantic schemas for Inspections and Inspection Templates."""
from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict


# ============================================================
# TEMPLATES
# ============================================================

class ChecklistItemCreate(BaseModel):
    label: str
    required: bool = True
    order: int = 0


class InspectionTemplateCreate(BaseModel):
    name: str
    description: Optional[str] = None
    category: str = "GENERAL"
    checklist_items: list[ChecklistItemCreate] = []


class InspectionTemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    checklist_items: Optional[list[ChecklistItemCreate]] = None


class InspectionTemplateResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    organization_id: UUID
    name: str
    description: Optional[str] = None
    category: str
    checklist_items: list[dict] = []
    is_default: bool = False
    created_at: datetime
    updated_at: datetime


class InspectionTemplateListResponse(BaseModel):
    data: list[InspectionTemplateResponse]
    meta: dict


# ============================================================
# INSPECTIONS
# ============================================================

class InspectionCreate(BaseModel):
    template_id: Optional[UUID] = None
    title: Optional[str] = None
    category: str = "GENERAL"
    scheduled_date: Optional[date] = None
    scheduled_time: Optional[str] = None
    location: Optional[str] = None
    inspector_name: Optional[str] = None
    inspector_company: Optional[str] = None
    notes: Optional[str] = None


class InspectionUpdate(BaseModel):
    title: Optional[str] = None
    category: Optional[str] = None
    scheduled_date: Optional[date] = None
    scheduled_time: Optional[str] = None
    location: Optional[str] = None
    inspector_name: Optional[str] = None
    inspector_company: Optional[str] = None
    notes: Optional[str] = None


class ChecklistResult(BaseModel):
    item_label: str
    result: str  # PASS | FAIL | NA
    notes: Optional[str] = None


class InspectionResultSubmit(BaseModel):
    results: list[ChecklistResult]
    overall_result: str  # PASSED | FAILED | CONDITIONAL
    notes: Optional[str] = None


class InspectionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    project_id: UUID
    number: int
    formatted_number: str
    title: str
    template_id: Optional[UUID] = None
    template_name: Optional[str] = None
    category: str
    scheduled_date: Optional[date] = None
    scheduled_time: Optional[str] = None
    location: Optional[str] = None
    inspector_name: Optional[str] = None
    inspector_company: Optional[str] = None
    status: str
    overall_result: Optional[str] = None
    checklist_results: list[dict] = []
    photo_ids: list = []
    notes: Optional[str] = None
    comments_count: int = 0
    created_by: Optional[UUID] = None
    created_by_name: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    updated_at: datetime


class InspectionListResponse(BaseModel):
    data: list[InspectionResponse]
    meta: dict

File: apps/web/src/types/inspection.ts
export interface InspectionTemplate {
  id: string;
  organization_id: string;
  name: string;
  description?: string | null;
  category: string;
  checklist_items: ChecklistItem[];
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface ChecklistItem {
  label: string;
  required: boolean;
  order: number;
}

export interface Inspection {
  id: string;
  project_id: string;
  number: number;
  formatted_number: string;
  title: string;
  template_id?: string | null;
  template_name?: string | null;
  category: string;
  scheduled_date?: string | null;
  scheduled_time?: string | null;
  location?: string | null;
  inspector_name?: string | null;
  inspector_company?: string | null;
  status: string;
  overall_result?: string | null;
  checklist_results: ChecklistResult[];
  photo_ids: string[];
  notes?: string | null;
  comments_count: number;
  created_by?: string | null;
  created_by_name?: string | null;
  created_at: string;
  completed_at?: string | null;
  updated_at: string;
}

export interface ChecklistResult {
  item_label: string;
  result: string;
  notes?: string | null;
}

export interface InspectionCreateInput {
  template_id?: string;
  title?: string;
  category?: string;
  scheduled_date?: string;
  scheduled_time?: string;
  location?: string;
  inspector_name?: string;
  inspector_company?: string;
  notes?: string;
}
